IPython/terminal/prompts.py

```diff
@@ -62,11 +62,15 @@ class RichPromptDisplayHook(DisplayHook):
     """Subclass of base display hook using coloured prompt"""
     def write_output_prompt(self):
         sys.stdout.write(self.shell.separate_out)
-        self.prompt_end_newline = False
+        # If we're not displaying a prompt, it effectively ends with a newline,
+        # because the output will be left-aligned.
+        self.prompt_end_newline = True
+
         if self.do_full_cache:
             tokens = self.shell.prompts.out_prompt_tokens()
-            if tokens and tokens[-1][1].endswith('\n'):
-                self.prompt_end_newline = True
+            if tokens and not tokens[-1][1].endswith('\n'):
```

Comment:
The test should be
 `tokens and len(tokens[-1][1]) > 0 and not tokens[-1][1].endswith('\n')`